src/probability/upcrossing.lean

```diff
@@ -0,0 +1,914 @@
+  (hf : submartingale f ℱ μ) (hfN : ∀ x, a ≤ f N x) (hfzero : 0 ≤ f 0) (hN : 0 < N) (hab : a < b) :
+  (b - a) * μ[upcrossings_before a b f N] ≤ μ[f N] :=
+calc (b - a) * μ[upcrossings_before a b f N]
+     ≤ μ[∑ k in finset.range N, upcrossing_strat a b f N k * (f (k + 1) - f k)] :
+begin
+  rw ← integral_mul_left,
+  refine integral_mono_of_nonneg _ ((hf.sum_upcrossing_strat_mul a b N).integrable N) _,
+  { exact eventually_of_forall (λ x, mul_nonneg (sub_nonneg.2 hab.le) (nat.cast_nonneg _)) },
+  { refine eventually_of_forall (λ x, _),
+    simpa using mul_upcrossings_before_le (hfN x) hN hab },
+end
+  ...≤ μ[f N] - μ[f 0] : hf.sum_mul_upcrossing_strat_le
+  ...≤ μ[f N] : (sub_le_self_iff _).2 (integral_nonneg hfzero)
+
+lemma crossing_pos_eq (hab : a < b) :
+  upper_crossing_time 0 (b - a) (λ n x, (f n x - a)⁺) N n = upper_crossing_time a b f N n ∧
+  lower_crossing_time 0 (b - a) (λ n x, (f n x - a)⁺) N n = lower_crossing_time a b f N n :=
+begin
+  have hab' : 0 < b - a := sub_pos.2 hab,
+  have hf : ∀ x i, b - a ≤ (f i x - a)⁺ ↔ b ≤ f i x,
+  { intros i x,
+    refine ⟨λ h, _, λ h, _⟩,
+    { rwa [← sub_le_sub_iff_right a,
+        ← lattice_ordered_comm_group.pos_of_pos_pos (lt_of_lt_of_le hab' h)] },
+    { rw ← sub_le_sub_iff_right a at h,
+      rwa lattice_ordered_comm_group.pos_of_nonneg _ (le_trans hab'.le h) } },
+  have hf' : ∀ x i, (f i x - a)⁺ ≤ 0 ↔ f i x ≤ a,
+  { intros x i,
+    rw [lattice_ordered_comm_group.pos_nonpos_iff, sub_nonpos] },
+  induction n with k ih,
+  { refine ⟨rfl, _⟩,
+    simp only [lower_crossing_time_zero, hitting, set.mem_Icc, set.mem_Iic],
+    ext x,
+    split_ifs with h₁ h₂ h₂,
+    { simp_rw [hf'] },
+    { simp_rw [set.mem_Iic, ← hf' _ _] at h₂,
+      exact false.elim (h₂ h₁) },
+    { simp_rw [set.mem_Iic, hf' _ _] at h₁,
+      exact false.elim (h₁ h₂) },
+    { refl } },
+  { have : upper_crossing_time 0 (b - a) (λ n x, (f n x - a)⁺) N (k + 1) =
+      upper_crossing_time a b f N (k + 1),
+    { ext x,
+      simp only [upper_crossing_time_succ_eq, ← ih.2, hitting, set.mem_Ici, tsub_le_iff_right],
+      split_ifs with h₁ h₂ h₂,
+      { simp_rw [← sub_le_iff_le_add, hf x] },
+      { simp_rw [set.mem_Ici, ← hf _ _] at h₂,
+        exact false.elim (h₂ h₁) },
+      { simp_rw [set.mem_Ici, hf _ _] at h₁,
+        exact false.elim (h₁ h₂) },
+      { refl } },
+    refine ⟨this, _⟩,
+    ext x,
+    simp only [lower_crossing_time, this, hitting, set.mem_Iic],
+    split_ifs with h₁ h₂ h₂,
+    { simp_rw [hf' x] },
+    { simp_rw [set.mem_Iic, ← hf' _ _] at h₂,
+      exact false.elim (h₂ h₁) },
+    { simp_rw [set.mem_Iic, hf' _ _] at h₁,
+      exact false.elim (h₁ h₂) },
+    { refl } }
+end
+
+lemma upcrossings_before_pos_eq (hab : a < b) :
+  upcrossings_before 0 (b - a) (λ n x, (f n x - a)⁺) N x = upcrossings_before a b f N x :=
+by simp_rw [upcrossings_before, (crossing_pos_eq hab).1]
+
+private lemma mul_integral_upcrossings_before_le_integral_pos_part'' [is_finite_measure μ]
```

Comment:
In general, we avoid private lemmas (it breaks follow to source and some docstrings). Also, primes do not indicate clearly if this is an auxiliary lemma, so postfixing the lemma name with `aux1` or `aux2` is better.